File: tui/api/chains.py
"""Chain and workflow domain client for the unified TUI."""

from __future__ import annotations

from .base import APIResult, DomainClient


class ChainsAPI(DomainClient):
    async def list_chains(self) -> APIResult:
        return await self.transport.get("/api/chains/")

    async def get_chain(self, chain_id: int) -> APIResult:
        return await self.transport.get(f"/api/chains/{chain_id}")

    async def create_chain(self, name: str) -> APIResult:
        return await self.transport.post("/api/chains/", json={"name": name})

    async def activate_chain(self, chain_id: int) -> APIResult:
        return await self.transport.post(f"/api/chains/{chain_id}/activate")

    async def toggle_plugin_bypass(
        self,
        chain_id: int,
        plugin_uri: str,
        *,
        bypass: bool,
        plugin_position: int | None = None,
    ) -> APIResult:
        params = {"bypass": str(bypass).lower()}
        if plugin_position is not None:
            params["plugin_position"] = plugin_position
        return await self.transport.post(
            f"/api/chains/{chain_id}/plugins/{plugin_uri}/bypass",
            params=params,
        )

    async def save_chain_preset(self, chain_id: int, preset_name: str) -> APIResult:
        return await self.transport.post(
            f"/api/chains/{chain_id}/preset/save",
            params={"preset_name": preset_name},
        )

    async def get_touchscreen_state(self, chain_id: int) -> APIResult:
        return await self.transport.get(f"/api/chains/{chain_id}/touchscreen")

    async def update_touchscreen_stomps(self, chain_id: int, assignments: list[dict]) -> APIResult:
        return await self.transport.put(
            f"/api/chains/{chain_id}/touchscreen/stomps",
            json={"assignments": assignments},
        )

    async def rename_chain(self, chain_id: int, new_name: str) -> APIResult:
        return await self.transport.put(
            f"/api/chains/{chain_id}/rename",
            params={"new_name": new_name},
        )

    async def delete_chain(self, chain_id: int) -> APIResult:
        return await self.transport.delete(f"/api/chains/{chain_id}")

    async def list_templates(self) -> APIResult:
        return await self.transport.get("/api/chains/templates/list")

    async def load_template(self, template_name: str) -> APIResult:
        return await self.transport.post(
            "/api/chains/templates/load",
            params={"template_name": template_name},
        )

    async def list_sessions(self) -> APIResult:
        return await self.transport.get("/api/sessions/list")

    async def save_session(self, name: str, description: str = "") -> APIResult:
        return await self.transport.post(
            "/api/sessions/save",
            json={"name": name, "description": description, "tags": []},
        )

    async def load_session(self, session_id: int) -> APIResult:
        return await self.transport.post(f"/api/sessions/load/{session_id}")

File: web/src/map2/clients/chains.ts
import type {
  Chain,
  ChainTemplate,
  ChainsResponse,
  PluginOrderRef,
  Snapshot,
} from '../types'
import { appendNodeQuery, fetchJson } from '../http'
import { API_BASE } from '../transport'
import {
  getChainReorderCompatibilityKey,
  getLegacyUriOnlyPluginOrder,
  isLegacyUriOnlyReorderValidationError,
  normalizeReorderPluginsResponse,
  type RawReorderPluginsResponse,
} from '../reorderPluginsCompat'

const legacyUriOnlyChainReorderNodes = new Set<string>()

async function postLegacyUriOnlyChainReorder(
  url: string,
  pluginOrder: PluginOrderRef[],
): Promise<{ status: string; chain_id: number; plugins: PluginOrderRef[] }> {
  const response = await fetchJson<RawReorderPluginsResponse>(url, {
    method: 'POST',
    body: JSON.stringify(getLegacyUriOnlyPluginOrder(pluginOrder)),
  })

  return normalizeReorderPluginsResponse(response, pluginOrder)
}

export const chainsApi = {
  list: (nodeId?: string | null) =>
    fetchJson<ChainsResponse>(appendNodeQuery(`${API_BASE}/chains/`, nodeId), { cache: 'no-store' }),

  get: (chainId: number, nodeId?: string | null) =>
    fetchJson<Chain>(appendNodeQuery(`${API_BASE}/chains/${chainId}`, nodeId), { cache: 'no-store' }),

  create: (name: string, nodeId?: string | null) =>
    fetchJson<Chain>(appendNodeQuery(`${API_BASE}/chains/`, nodeId), {
      method: 'POST',
      body: JSON.stringify({ name }),
    }),

  delete: (chainId: number, nodeId?: string | null) =>
    fetchJson<{ status: string; chain_id: number }>(appendNodeQuery(`${API_BASE}/chains/${chainId}`, nodeId), {
      method: 'DELETE',
    }),

  rename: (chainId: number, newName: string, nodeId?: string | null) =>
    fetchJson<{ status: string; chain_id: number; name: string }>(
      appendNodeQuery(`${API_BASE}/chains/${chainId}/rename?new_name=${encodeURIComponent(newName)}`, nodeId),
      { method: 'PUT' },
    ),

  activate: (chainId: number, nodeId?: string | null) =>
    fetchJson<{ status: string; chain_id: number }>(
      appendNodeQuery(`${API_BASE}/chains/${chainId}/activate`, nodeId),
      { method: 'POST' },
    ),

  deactivate: (chainId: number, nodeId?: string | null) =>
    fetchJson<{ status: string; chain_id: number }>(
      appendNodeQuery(`${API_BASE}/chains/${chainId}/deactivate`, nodeId),
      { method: 'POST' },
    ),

  addPlugin: (chainId: number, pluginUri: string, nodeId?: string | null) =>
    fetchJson<{ status: string; chain_id: number; plugin: string; plugins_count: number; plugin_position?: number }>(
      appendNodeQuery(`${API_BASE}/chains/${chainId}/plugins?plugin_uri=${encodeURIComponent(pluginUri)}`, nodeId),
      { method: 'POST' },
    ),

  removePlugin: (chainId: number, pluginUri: string, pluginPosition?: number, nodeId?: string | null) => {
    const params = new URLSearchParams({ plugin_uri: pluginUri })
    if (typeof pluginPosition === 'number' && Number.isFinite(pluginPosition)) {
      params.set('plugin_position', String(pluginPosition))
    }
    return fetchJson<{ status: string; chain_id: number }>(
      appendNodeQuery(`${API_BASE}/chains/${chainId}/plugins?${params.toString()}`, nodeId),
      { method: 'DELETE' },
    )
  },

  reorderPlugins: async (chainId: number, pluginOrder: PluginOrderRef[], nodeId?: string | null) => {
    const url = appendNodeQuery(`${API_BASE}/chains/${chainId}/reorder`, nodeId)
    const compatibilityKey = getChainReorderCompatibilityKey(nodeId)

    if (legacyUriOnlyChainReorderNodes.has(compatibilityKey)) {
      return postLegacyUriOnlyChainReorder(url, pluginOrder)
    }

    try {
      const response = await fetchJson<RawReorderPluginsResponse>(url, {
        method: 'POST',
        body: JSON.stringify(pluginOrder),
      })
      return normalizeReorderPluginsResponse(response, pluginOrder)
    } catch (error) {
      if (!isLegacyUriOnlyReorderValidationError(error)) {
        throw error
      }

      legacyUriOnlyChainReorderNodes.add(compatibilityKey)
      return postLegacyUriOnlyChainReorder(url, pluginOrder)
    }
  },

  togglePluginBypass: (chainId: number, pluginUri: string, bypass: boolean, pluginPosition?: number, nodeId?: string | null) => {
    const params = new URLSearchParams({ bypass: String(bypass) })
    if (typeof pluginPosition === 'number' && Number.isFinite(pluginPosition)) {
      params.set('plugin_position', String(pluginPosition))
    }
    return fetchJson<{ status: string; chain_id: number; plugin: string; bypass: boolean }>(
      appendNodeQuery(
        `${API_BASE}/chains/${chainId}/plugins/${encodeURIComponent(pluginUri)}/bypass?${params.toString()}`,
        nodeId,
      ),
      { method: 'POST' },
    )
  },

  savePreset: (chainId: number, presetName: string, nodeId?: string | null) =>
    fetchJson<{ status: string; preset_id: number; name: string }>(
      appendNodeQuery(`${API_BASE}/chains/${chainId}/preset/save?preset_name=${encodeURIComponent(presetName)}`, nodeId),
      { method: 'POST' },
    ),

  listPresets: (nodeId?: string | null) =>
    fetchJson<{ presets: Snapshot[]; count: number }>(appendNodeQuery(`${API_BASE}/chains/presets`, nodeId)),

  loadPreset: (presetId: number, nodeId?: string | null) =>
    fetchJson<{ status: string; chain_id: number }>(appendNodeQuery(`${API_BASE}/chains/preset/${presetId}/load`, nodeId), {
      method: 'POST',
    }),

  deletePreset: (presetId: number, nodeId?: string | null) =>
    fetchJson<{ status: string; preset_id: number }>(appendNodeQuery(`${API_BASE}/chains/preset/${presetId}`, nodeId), {
      method: 'DELETE',
    }),

  listTemplates: (nodeId?: string | null) =>
    fetchJson<{ templates: ChainTemplate[]; count: number }>(appendNodeQuery(`${API_BASE}/chains/templates/list`, nodeId)),

  loadTemplate: (templateName: string, nodeId?: string | null) =>
    fetchJson<{ status: string; chain: Chain }>(
      appendNodeQuery(`${API_BASE}/chains/templates/load?template_name=${encodeURIComponent(templateName)}`, nodeId),
      { method: 'POST' },
    ),
}
